packages/Discretization/src/DTK_Interpolation_def.hpp

```diff
@@ -0,0 +1,669 @@
+                for ( int j = offset( i ); j < offset( i + 1 ); ++j )
+                {
+                    exported_query_ids( j ) = i;
+                    for ( unsigned int k = 0; k < dim; ++k )
+                        exported_points( j )[k] = points_coord( i, k );
+                }
+            } );
+        Kokkos::fence();
+    }
+
+    // Communicate the points
+    Kokkos::realloc( imported_points, n_imports );
+    DataTransferKit::DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+        distributed_search_distributor, exported_points, imported_points );
+
+    // Communicate the query_ids. We communicate the query_ids to keep track of
+    // which points is associated to which query.
+    Kokkos::realloc( imported_query_ids, n_imports );
+    DataTransferKit::DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+        distributed_search_distributor, exported_query_ids,
+        imported_query_ids );
+
+    // We would like to get the reverse of the distributor but the following
+    // code hangs during the communication
+    //
+    // Tpetra::Distributor
+    // tmp_1(*(distributed_search_distributor.getReverse()));
+    // Tpetra::Distributor tmp_2(*(tmp_1.getReverse()));
+    // DataTransferKit::DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+    //     tmp_2, exported_points, imported_points );
+    //
+    // tmp_2 should be the same as distributed_search_distributor yet the code
+    // hangs. So instead, create a new distributor from scratch. Note that I
+    // cannot reproduce this simple on a simple test case.
+    Kokkos::View<int *, DeviceType> own_rank( "own_rank", ranks.extent( 0 ) );
+    Kokkos::deep_copy( own_rank, _comm->getRank() );
+    Kokkos::View<int *, DeviceType> imported_ranks( "imported_ranks",
+                                                    n_imports );
+    DataTransferKit::DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+        distributed_search_distributor, own_rank, imported_ranks );
+    auto imported_ranks_host = Kokkos::create_mirror_view( imported_ranks );
+    Kokkos::deep_copy( imported_ranks_host, imported_ranks );
+    _distributor.createFromSends( Teuchos::ArrayView<int const>(
+        imported_ranks_host.data(), imported_ranks_host.extent( 0 ) ) );
+}
+
+template <typename DeviceType>
+void Interpolation<DeviceType>::filterTopology(
+    Kokkos::View<unsigned int *, DeviceType> topo, unsigned int topo_id,
+    Kokkos::View<int *, DeviceType> cell_indices,
+    Kokkos::View<DataTransferKit::Point *, DeviceType> points,
+    Kokkos::View<int *, DeviceType> query_ids,
+    Kokkos::View<unsigned int *, DeviceType> map,
+    Kokkos::View<int *, DeviceType> filtered_cell_indices,
+    Kokkos::View<double **, DeviceType> filtered_points,
+    Kokkos::View<int *, DeviceType> filtered_query_ids )
+{
+    using ExecutionSpace = typename DeviceType::execution_space;
+    unsigned int const size = topo.extent( 0 );
```

Comment:
This is not a very helpful name.  Is that the total number of cells regardless of their topology?

Reply:
It's the size of n_imports -> total number of cells where we will do the the point in cell search.